pkg/pillar/pubsub/pubsubimpl.go

```diff
@@ -0,0 +1,246 @@
+// Copyright (c) 2017,2018 Zededa, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+// Provide for a pubsub mechanism for config and status which is
+// backed by an IPC mechanism such as connected sockets.
+
+package pubsub
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"os"
+	"path"
+
+	log "github.com/sirupsen/logrus"
+)
+
+// Protocol over AF_UNIX or other IPC mechanism
+// "request" from client after connect to sanity check subject.
+// Server sends the other messages; "update" for initial values.
+// "complete" once all initial keys/values in collection have been sent.
+// "restarted" if/when pub.km.restarted is set.
+// Ongoing we send "update" and "delete" messages.
+// They keys and values are base64-encoded since they might contain spaces.
+// We include typeName after command word for sanity checks.
+// Hence the message format is
+//	"request" topic
+//	"hello"  topic
+//	"update" topic key json-val
+//	"delete" topic key
+//	"complete" topic (aka synchronized)
+//	"restarted" topic
+
+// We always publish to our collection.
+// We always write to a file in order to have a checkpoint on restart
```

Comment:
We may not want to always write to a file.. 
For example, when communicating Username / Passwords between the Decryptor and the application, the communication can be through PubSub, but it doesn't need to get written to a file. The restart semantics would be that the application would start over again..

Reply:
This comment was copied over from the current implementation. I assumed it meant that the implementation does these writes. In any case, I didn't change this logic or the comments, just ported over to `PubSubImpl`.